### PR TITLE
Add service tests

### PR DESCRIPTION
**Pull request by trauman413**
## Jira Ticket Number
https://gaming-backlog.atlassian.net/browse/GB-79

## Short description
Include a 1-3 sentence description of what is being done (doesn't need to be fancy)
Adds better unit tests to actually test the services rather than run tests on the controllers. This provides testing at a more granular level.

## Type of Change
- [ ] New feature
- [ ] Bug fix
- [ ] Documentation
- [x] Other (please specify): Testing

### CODE REVIEW

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt [hunk 1]
```diff
@@ -0,0 +1,69 @@
...
+  private val mockIgdbClient = Mockito.mock(IGDBClient::class.java)
+  private val mockGameRepository = Mockito.mock(GameRepository::class.java)
+
+  private val gameService = GameService(mockGameRepository, mockIgdbClient)
+
+  @Nested
+  @DisplayName("Tests igdbToGame")
+  inner class IGDBToGameTests {
+    @Test
+    fun shouldSuccessfullyConvertNullFranchise() {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt:28]: We're just testing here that it creates a game successfully? We don't actually do anything with the franchise?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+      val result = gameService.igdbGameToGame(igdbGame)
+      assertEquals("12", result.igdbId)
+      assertEquals("Best game ever", result.summary)
+      assertEquals(listOf("Platform", "RPG"), result.genres)
+      assertEquals("ABC", result.name)
+      assertEquals(listOf("Nintendo Switch"), result.platforms)
+    }
+
+    @Test
+    fun shouldSuccessfullyConvert() {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt]: Can we be a little more specific here? `shouldSuccessfullyConvertWithAllFields`?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt [hunk 3]
```diff
@@ -1,4 +1,116 @@
...
+    }
+
+    @Test
+    fun shouldCreateLibraryWithGames() {
+      val library = Library(libraryId, "Backlog", arrayListOf(gameId1))
+      mockSave(library)
+      val request = LibraryRequest("Backlog", arrayListOf(gameId1))
+      val result = libraryService.create(request)
+      assertEquals(library.name, result.name)
+      assertEquals(library.games, result.games)
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: This is identical to the test above except for the games. Can we make a utility for this and just pass in the params for the library that we want?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt [hunk 4]
```diff
@@ -1,4 +1,116 @@
...
+
+  @Nested
+  @DisplayName("Tests for deleteGameFromLibrary")
+  inner class DeleteGameFromLibrary {
+    @Test
+    fun shouldDeleteGameFromLibrary() {
+      mockLibraryDb[libraryId] = Library(libraryId, "Backlog", arrayListOf(gameId1))
+      mockGetOne(libraryId)
+      mockLibraryDb[libraryId]?.games?.remove(gameId1)
+      mockLibraryDb[libraryId]?.let { mockSave(it) }
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: Can we make a helper method for this as well (since it's used here and in the add test)

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt [hunk 5]
```diff
@@ -1,4 +1,116 @@
...
+      val result = libraryService.deleteGameFromLibrary(libraryId, gameId1)
+      assertEquals(mockLibraryDb[libraryId], result)
+    }
+
+    @Test
+    fun shouldNotDeleteIfNoIDIsFound() {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt:111]: Do we also want a test for when the library exists but the game doesn't?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt [hunk 6]
```diff
@@ -87,6 +85,7 @@ class UserServiceTests {
       assertEquals(user.libraries, result?.libraries)
     }
 
+    // No tests for if fields are null because they can't actually be null
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt]: Is that validated in the constructor?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt [hunk 7]
```diff
@@ -141,4 +140,40 @@ class UserServiceTests {
...
+      val updateRequest = UserRequest("user2", "456", "newtest@test.com")
+      val result = userService.updateUserFields(ogUser.id, updateRequest)
+      assertEquals(mockUserDb[ogUser.id]?.displayName, result?.displayName)
+      assertEquals(mockUserDb[ogUser.id]?.email, result?.email)
+      assertEquals(mockUserDb[ogUser.id]?.password, result?.password)
+      assertEquals(mockUserDb[ogUser.id]?.libraries, result?.libraries)
+    }
+
+    @Test
+    fun shouldUpdatePartialUserFields() {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt:177]: Same idea here with extracting duplicated logic

**Review by ksadoff — COMMENTED** [review]
Mainly just comments about refactoring duplicate logic, otherwise looks good

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt (hunk 1 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt:28]: Yep: this is accounting for a previous error we had when we weren't accounting for null franchises and returned something weird. Nothing is done specifically but it's still an edge case worth accounting for.

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 4 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: now that i'm revisiting these tests several weeks later, i don't think they're actually doing anything. Actual DB tests weren't in scope for this ticket but those will actually test this functionality. I'm tempted to remove the add to library/delete from library tests since I think things are being too mocked out to actually test any functionality. Thoughts?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 5 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt:111]: The actual code isn't checking if the gameId exists. That might be an oversight in the actual code, and I don't think mocking things will accurately get that case to see for sure right now.

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 4 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: I fixed this up anyway and will push shortly but worth considering if it's necessary

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt (hunk 6 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt]: i'm gonna be honest, i have no idea what i was talking about three weeks ago 😅 i added some tests here lol

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt [hunk 8]
```diff
@@ -49,27 +49,33 @@ class LibraryServiceTests {
...
+  }
+
   @Nested
   @DisplayName("Tests for createLibrary")
   inner class CreateLibrary {
-    @Test
-    fun shouldCreateLibraryWithNoGames() {
-      val library = Library(libraryId, "Backlog", arrayListOf())
+
+    private fun testCreateLibraryUtil(libraryId: String, libraryName: String, gameIds: ArrayList<String>) {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt:62]: [Nit] Do we need the "test" prefix? That tends to make me think of an actual test function as opposed to a helper

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 5 above)
Comment (ksadoff) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt:111]: Should we add that validation in? Can do that and the corresponding test in another ticket

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt [hunk 9]
```diff
@@ -0,0 +1,69 @@
+package com.gamingbacklog.api.gamingbacklogapi.unit.services
+
+import com.gamingbacklog.api.gamingbacklogapi.clients.IGDBClient
+import com.gamingbacklog.api.gamingbacklogapi.models.igdb.*
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/GameServiceTests.kt]: [Nit] Can we avoid using * imports?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt [hunk 10]
```diff
@@ -1,4 +1,117 @@
...
+      val library = Library(libraryId, libraryName, gameIds)
+      val request = LibraryRequest(libraryName, gameIds)
+      mockSave(library)
+      val result = libraryService.create(request)
+      assertEquals(library.name, result.name)
+      assertEquals(library.games, result.games)
+
+    }
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt:69]: [Nit] spacing

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt [hunk 11]
```diff
@@ -1,4 +1,117 @@
...
+    @Test
+    fun shouldAddToLibrary() {
+      saveMockLibraryDB(libraryId, "Backlog", arrayListOf())
+      mockLibraryDb[libraryId]?.games?.add(gameId1)
+      val result = libraryService.addToLibrary(libraryId, gameId1)
+      assertEquals(mockLibraryDb[libraryId], result)
+    }
+
+    @Test
+    fun shouldAddToLibraryNoIDFound() {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: ?
```suggestion
    fun shouldNotAddToLibraryNoIDFound() {
```

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 4 above)
Comment (ksadoff) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: Discussed offline: we can remove these tests

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt [hunk 12]
```diff
@@ -141,4 +150,44 @@ class UserServiceTests {
...
+      val updateRequest = UserRequest(displayName = "user2", password = null, email = "newtest@test.com")
+      val result = userService.updateUserFields(ogUser.id, updateRequest)
+      assertEquals(mockUserDb[ogUser.id]?.displayName, result?.displayName)
+      assertEquals(mockUserDb[ogUser.id]?.email, result?.email)
+      assertEquals(mockUserDb[ogUser.id]?.password, result?.password)
+      assertEquals(mockUserDb[ogUser.id]?.libraries, result?.libraries)
+    }
+
+    @Test
+    fun shouldFindNoUser() {
```
Comment (ksadoff) [inline src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt]: `shouldFindNoUser` when what?

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 5 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt:111]: Created https://gaming-backlog.atlassian.net/browse/GB-89 (though I guess we don't need the corresponding test since we removed the pointless tests as per the earlier comment. we'll account for that when we actually test repo logic)

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 11 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt:94]: you're right, but removed the test anyway :)

Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: I think I was calling it as such because it's an Add To Library test with the case No ID is found but I'll change it to your suggestion (and a bonus "If") to make it clearer

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 4 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: actually may leave them for now just to have some coverage and because I think user tests have the same issue and I don't want to deal with it there :P (even if they're completely useless). I'll remove them when we write repository tests (no ticket yet, will create after discussion on discord)

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt (hunk 12 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/UserServiceTests.kt]: fixed

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 11 above)
Comment (ksadoff) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: Ah okay, in that case in the future we can probably use underscores to make the parsing more clear

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 4 above)
Comment (ksadoff) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: Is there a ticket now? Would be good to link to this comment thread in the ticket

**Review by ksadoff — APPROVED** [review]
LGTM with two last comments!

File: src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt (hunk 4 above)
Comment (trauman413) [reply src/test/kotlin/com/gamingbacklog/api/gamingbacklogapi/unit/services/LibraryServiceTests.kt]: Added comment to https://gaming-backlog.atlassian.net/browse/GB-47
